opencog/atoms/bind/BindLink.cc

```diff
@@ -0,0 +1,107 @@
+/*
+ * BindLink.cc
+ *
+ * Copyright (C) 2009, 2014, 2015 Linas Vepstas
+ *
+ * Author: Linas Vepstas <linasvepstas@gmail.com>  January 2009
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License v3 as
+ * published by the Free Software Foundation and including the
+ * exceptions
+ * at http://opencog.org/wiki/Licenses
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public
+ * License
+ * along with this program; if not, write to:
+ * Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include <opencog/atomspace/ClassServer.h>
+#include <opencog/query/PatternMatch.h>
+
+#include "BindLink.h"
+
+using namespace opencog;
+
+BindLink::BindLink(Type t, const HandleSeq& hseq,
+                   TruthValuePtr tv, AttentionValuePtr av)
+	: SatisfactionLink(t, hseq, tv, av)
+{
+	if (not classserver().isA(t, BIND_LINK))
+	{
+		const std::string& tname = classserver().getTypeName(t);
+		throw InvalidParamException(TRACE_INFO,
+			"Expecting a BindLink, got %s", tname.c_str());
+	}
+
+	// Don't throw an error, don't initialize if its the pattern matcher
+	if (NULL == _body) return;  // XXX temp hack ??
+
+	// The LambdaLink contructor sets up _body and _varset
+	validate_body(_body);
+	unbundle_clauses(_hclauses);
+	validate_clauses(_varset, _clauses);
+}
+
+BindLink::BindLink(Link &l)
+	: SatisfactionLink(l)
+{
+	Type t = l.getType();
+	if (not classserver().isA(t, BIND_LINK))
+	{
+		const std::string& tname = classserver().getTypeName(t);
+		throw InvalidParamException(TRACE_INFO,
+			"Expecting a BindLink, got %s", tname.c_str());
+	}
+
+	// The LambdaLink contructor sets up _body and _varset
+	validate_body(_body);
+	unbundle_clauses(_hclauses);
+	validate_clauses(_varset, _clauses);
```

Comment:
Hello Linas,

The early checking of BindLink does not allow me to create the complete BindLink outside the AtomSpace, and then adding the whole thing all at once to the AtomSpace, since the validation code are comparing Handles, which are UNDEFINED until everything is in the AtomSpace.

I could work around this in my code by adding things to AtomSpace immediately as I build.   Just thought you should know.